cf-custom-resources/lib/custom-domain-app-runner.js

```diff
@@ -0,0 +1,189 @@
+            console.log("Custom domain already associated. Do nothing.");
+            return;
+        }
+        console.log(`Caught error: ${err.message}`);
+        await report(event, context, "FAILED", event.LogicalResourceId, null, err.message).catch((err) => {
+            throw new Error("send response: " + err.message);
+        });
+    });
+
+    if (!addCustomDomainErr) {
+        console.log("Finished");
+        await report(event, context, "SUCCESS", event.LogicalResourceId).catch((err) => {
+            throw new Error("send response: " + err.message);
+        });
+    }
+};
+
+/**
+ * Validate certificates of the custom domain for the service by upserting validation records.
+ * Errors are not handled and are directly passed to the caller.
+ *
+ * @param {string} serviceARN ARN of the service that the custom domain applies to.
+ * @param {string} customDomainName the custom domain name.
+ */
+async function addCustomDomain(serviceARN, customDomainName) {
+    const data = await appRunnerClient.associateCustomDomain({
+        DomainName: customDomainName,
+        ServiceArn: serviceARN,
+    }).promise().catch(err => {
+        if (err.name === ERR_NAME_INVALID_REQUEST && err.message.includes(`${customDomainName} is already associated with`)) {
+            throw new CustomDomainError(`${customDomainName} is already associated with service ${serviceARN}`, ERR_NAME_CUSTOM_DOMAIN_ALREADY_ASSOCIATED);
+        }
+        throw err;
+    });
+
+    await upsertCNAMERecordAndWait(customDomainName, data.DNSTarget, appHostedZoneID);
+    await validateCertForDomain(serviceARN, customDomainName);
+}
+
+/**
+ * Validate certificates of the custom domain for the service by upserting validation records.
+ *
+ * @param {string} serviceARN ARN of the service that the custom domain applies to.
+ * @param {string} domainName the custom domain name.
+ * @throws wrapped error.
+ */
+async function validateCertForDomain(serviceARN, domainName) {
+    console.log("Add validation records");
+    let i;
+    for (i = 0; i < ATTEMPTS; i++){
+        const data = await appRunnerClient.describeCustomDomains({
+            ServiceArn: serviceARN,
+        }).promise().catch(err => {
+            throw new Error(`get custom domains for service ${serviceARN}: ` + err.message);
+        });
+
+        const customDomains = data.CustomDomains;
+        let domain;
+        for (const i in customDomains) {
+            if (customDomains[i].DomainName === domainName) {
+                domain = customDomains[i];
+                break;
+            }
+        }
+        if (!domain) {
+            throw new Error(`domain ${domainName} is not associated`);
+        }
+
+        if (domain.Status !== DOMAIN_STATUS_PENDING_VERIFICATION) {
+            console.log(`Custom domain ${domainName} status is ${domain.Status}. Desired state is ${DOMAIN_STATUS_PENDING_VERIFICATION}. Wait and check again`);
+            await sleep(3000);
```

Comment:
Instead we could do exponential backoff https://github.com/aws/copilot-cli/blob/4212955bf7b6238be408620061715e6f5472ccb5/cf-custom-resources/lib/dns-cert-validator.js#L159-L163

Reply:
I think it's worth doing exponential back off to wait for `ACTIVE`. The `PENDING` status though is usually quick.

Reply:
sgtm